cpp/src/strings/findall.cu

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/null_mask.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/char_types/char_types.hpp>
+#include <cudf/strings/findall.hpp>
+#include <strings/utilities.hpp>
+#include <strings/regex/regex.cuh>
+
+#include <thrust/extrema.h>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+using string_index_pair = thrust::pair<const char*,size_type>;
+
+namespace
+{
+
+/**
+ * @brief This functor handles counting the number of matches per string.
+ *
+ * @tparam stack_size Correlates to the number of regex instructions.
+ *         Each instruction requires a fixed sized memory overhead.
+ */
+template<size_t stack_size>
+struct findall_count_fn
+{
+    column_device_view const d_strings;
+    reprog_device prog;
+
+    __device__ size_type operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        u_char data1[stack_size], data2[stack_size];
+        prog.set_stack_mem(data1,data2);
+        string_view d_str = d_strings.element<string_view>(idx);
+        auto nchars = d_str.length();
+        size_type spos = 0, epos = nchars, column_count = 0;
+        while( spos <= nchars )
+        {
+            if( prog.find(idx,d_str,spos,epos) <=0 )
+                break; // no more matches found
+            spos = epos > spos ? epos : spos + 1;
+            epos = nchars;
+            ++column_count;
+        }
+        return column_count;
+    }
+};
+
+/**
+ * @brief This functor handles extracting matched strings by applying the compiled regex pattern
+ * and creating string_index_pairs for all the substrings.
+ */
+template<size_t stack_size>
+struct findall_fn
+{
+    column_device_view const d_strings;
+    reprog_device prog;
+    size_type column_index;
+    size_type const* d_counts;
+
+    __device__ string_index_pair operator()(size_type idx)
+    {
+        string_index_pair result{nullptr,0};
+        if( d_strings.is_null(idx) || (column_index >= d_counts[idx]) )
+            return result;
+        u_char data1[stack_size], data2[stack_size];
+        prog.set_stack_mem(data1,data2);
+        string_view d_str = d_strings.element<string_view>(idx);
+        auto nchars = d_str.length();
+        size_type spos = 0, epos = nchars, column_count = 0;
+        while( spos <= nchars )
+        {
+            if( prog.find(idx,d_str,spos,epos) <=0 )
+                break; // no more matches found
+            if( column_count == column_index )
+                break; // found our column
```

Comment:
There is a subset of the `findall_fn` and `finall_count_fn` code that is identical except for two lines:
```cpp
u_char data1[stack_size];
        u_char data2[stack_size];
        prog.set_stack_mem(data1,data2);
        string_view d_str = d_strings.element<string_view>(idx);
        auto nchars = d_str.length();
        size_type spos = 0, epos = nchars, column_count = 0;
        while( spos <= nchars )
        {
            if( prog.find(idx,d_str,spos,epos) <=0 )
                break; // no more matches found
            if( column_count == column_index )
                break; // found our column
            spos = epos > spos ? epos : spos + 1;
            epos = nchars;
            ++column_count;
        }
```
The only difference is
```cpp
            if( column_count == column_index )
                break; // found our column
```
Is it worth / possible to combine / factor out the common code? Seems a little redundant to have two copies essentially.